src/sources/postgresql_metrics.rs

```diff
@@ -182,6 +182,111 @@ impl SourceConfig for PostgresqlMetricsConfig {
     }
 }
 
+#[derive(Debug)]
+struct PostgresqlClient {
+    config: Config,
+    tls_config: Option<PostgresqlMetricsTlsConfig>,
+    client: Option<(Client, usize)>,
+}
+
+impl PostgresqlClient {
+    fn new(config: Config, tls_config: Option<PostgresqlMetricsTlsConfig>) -> Self {
+        Self {
+            config,
+            tls_config,
+            client: None,
+        }
+    }
+
+    async fn take(&mut self) -> Result<(Client, usize), ConnectError> {
+        match self.client.take() {
+            Some((client, version)) => Ok((client, version)),
+            None => self.build_client().await,
+        }
+    }
+
+    fn set(&mut self, value: (Client, usize)) {
+        self.client.replace(value);
+    }
+
+    async fn build_client(&self) -> Result<(Client, usize), ConnectError> {
+        // Create postgresql client
+        let client = match &self.tls_config {
+            Some(tls_config) => {
+                let mut builder =
+                    SslConnector::builder(SslMethod::tls_client()).context(TlsFailed)?;
+                builder
+                    .set_ca_file(tls_config.ca_file.clone())
+                    .context(TlsFailed)?;
+                let connector = MakeTlsConnector::new(builder.build());
+
+                let (client, connection) =
+                    self.config
+                        .connect(connector)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+            None => {
+                let (client, connection) =
+                    self.config
+                        .connect(NoTls)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+        };
+
+        // Log version if required
+        if tracing::level_enabled!(tracing::Level::DEBUG) {
+            let version_row = client
+                .query_one("SELECT version()", &[])
+                .await
+                .with_context(|| SelectVersionFailed {
+                    endpoint: config_to_endpoint(&self.config),
+                })?;
+            let version = version_row
+                .try_get::<&str, &str>("version")
+                .with_context(|| SelectVersionFailed {
+                    endpoint: config_to_endpoint(&self.config),
+                })?;
+            debug!(message = "Connected to server.", endpoint = %config_to_endpoint(&self.config), server_version = %version);
+        }
+
+        // Get server version and check that we support it
+        let row = client
```

Comment:
Is there a reason we need to query for the version twice (here and above for logging)? It seems like we could just query once unless I'm missing something. Sorry, I realize it was already like this, but it jumped out to me.

Reply:
While these queries return nearly the same, the second one much easier to parse:
```
vector=# select version();
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                     version                                                      │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ PostgreSQL 13.1 (Debian 13.1-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(1 row)

Time: 1.967 ms
vector=# show server_version_num;
┌────────────────────┐
│ server_version_num │
├────────────────────┤
│ 130001             │
└────────────────────┘
(1 row)

Time: 1.147 ms
```
also, we call query `select version()` only if `debug` (or lower) log level enabled.

Reply:
Yeah, I was just thinking that maybe we could just always execute the query, but only log at the debug level. I do see both are available via the catalogue:

```
postgres=# select * from pg_settings where name in ('server_version_num', 'server_version');
        name        |            setting             | unit |    category    |               short_desc                | extra_desc | context  | vartype | source  | min_val | max_val | enumvals |            boot_val            |           
reset_val            | sourcefile | sourceline | pending_restart 
--------------------+--------------------------------+------+----------------+-----------------------------------------+------------+----------+---------+---------+---------+---------+----------+--------------------------------+-----------
---------------------+------------+------------+-----------------
 server_version     | 13.1 (Debian 13.1-1.pgdg100+1) |      | Preset Options | Shows the server version.               |            | internal | string  | default |         |         |          | 13.1 (Debian 13.1-1.pgdg100+1) | 13.1 (Debi
an 13.1-1.pgdg100+1) |            |            | f
 server_version_num | 130001                         |      | Preset Options | Shows the server version as an integer. |            | internal | integer | default | 130001  | 130001  |          | 130001                         | 130001    
                     |            |            | f
(2 rows)
```

So it seems feasible to do it in one query, but I'm ok with it as-is given that the additional query only happens if debug is enabled.